relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -254,6 +260,112 @@ mod tests {
         envelope
     }
 
+    fn state_with_rule_and_condition(
+        sample_rate: Option<f64>,
+        rule_type: RuleType,
+        mode: SamplingMode,
+        condition: RuleCondition,
+    ) -> ProjectState {
+        let rules = match sample_rate {
+            Some(sample_rate) => vec![SamplingRule {
+                condition,
+                sample_rate,
+                ty: rule_type,
+                id: RuleId(1),
+                time_range: Default::default(),
+            }],
+            None => Vec::new(),
+        };
+
+        state_with_config(SamplingConfig {
+            rules,
+            mode,
+            next_id: None,
+        })
+    }
+
+    ///
+    ///
+    #[test]
+    fn test_trace_rules_applied_after_event_rules() {
+        // a transaction rule that drops everything
+        let event_state = state_with_rule_and_condition(
+            Some(0.0),
+            RuleType::Transaction,
+            SamplingMode::Received,
+            RuleCondition::Eq(EqCondition {
+                name: "event.transaction".to_owned(),
+                value: "healthcheck".into(),
+                options: Default::default(),
+            }),
+        );
+
+        // a trace rule that keeps everything
+        let trace_state = state_with_rule(Some(1.0), RuleType::Trace, SamplingMode::Received);
+
+        let matching_envelope = new_envelope(true, "healthcheck");
+        let non_matching_envelope = new_envelope(true, "test1");
```

Comment:
`matching` and `non_matching` are ambiguous, because both envelopes match exactly one of the two rules. So I would rename to something like:

```suggestion
        let healthcheck_envelope = new_envelope(true, "healthcheck");
        let other_envelope = new_envelope(true, "test1");
```

Same goes for the events below.

Reply:
good idea, I've changed it in my latest commit